clang/runtime/dpct-rt/include/dpl_extras/algorithm.h

```diff
@@ -1455,8 +1455,604 @@ inline void segmented_sort_pairs_by_two_pair_sorts(
+                                           __self_item);
+        });
+  });
+  e.wait();
+}
+
+template <::std::uint16_t WorkGroupSize, ::std::uint16_t ItersPerWorkItem,
+          typename BinType, typename Policy, typename Range1, typename Range2,
+          typename Size, typename IdxHashFunc, typename... Range3>
+inline void
+__histogram_general_local_atomics(Policy &&policy, Range1 &&input,
+                                  Range2 &&bins, const Size &num_bins,
+                                  IdxHashFunc func, Range3 &&...opt_range) {
+  const ::std::size_t N = input.size();
+  std::size_t segments = __ceiling_div(N, WorkGroupSize * ItersPerWorkItem);
+  auto e = policy.queue().submit([&](auto &h) {
+    // Temporary use of stable non-public API from oneDPL,  this function will
+    // be replaced with oneDPL call in an upcoming release.
+    oneapi::dpl::__ranges::__require_access(h, input, bins, opt_range...);
+    sycl::local_accessor<::std::uint32_t, 1> local_histogram(
+        sycl::range(num_bins), h);
+    h.parallel_for(
+        sycl::nd_range<1>(segments * WorkGroupSize, WorkGroupSize),
+        [=](sycl::nd_item<1> __self_item) {
+          const ::std::size_t __self_lidx = __self_item.get_local_id(0);
+          const ::std::uint32_t __wgroup_idx = __self_item.get_group(0);
+          const ::std::size_t __seg_start =
+              WorkGroupSize * __wgroup_idx * ItersPerWorkItem;
+
+          __clear_wglocal_histograms(local_histogram, 0, num_bins, __self_item);
+
+          __accum_local_atomics<WorkGroupSize, ::std::uint16_t,
+                                sycl::access::address_space::local_space>(
+              input, __seg_start, N, __self_lidx, local_histogram, 0, func,
+              __self_item, ItersPerWorkItem);
+
+          __reduce_out_histograms<BinType>(local_histogram, 0, bins, num_bins,
+                                           __self_item);
+        });
+  });
+
+  e.wait();
+}
+
+template <::std::uint16_t WorkGroupSize,
+          ::std::uint16_t __min_iters_per_work_item, typename BinType,
+          typename Policy, typename Range1, typename Range2, typename Size,
+          typename IdxHashFunc, typename... Range3>
+inline void __histogram_general_private_global_atomics(
+    Policy &&policy, Range1 &&input, Range2 &&bins, const Size &num_bins,
+    IdxHashFunc func, Range3 &&...opt_range) {
+
+  const ::std::size_t N = input.size();
+  auto __global_mem_size =
+      policy.queue()
+          .get_device()
+          .template get_info<sycl::info::device::global_mem_size>();
+  const ::std::size_t max_segments =
+      ::std::min(__global_mem_size / (num_bins * sizeof(BinType)),
+                 __ceiling_div(N, WorkGroupSize * __min_iters_per_work_item));
+  const ::std::size_t iters_per_work_item =
+      __ceiling_div(N, max_segments * WorkGroupSize);
+  ::std::size_t segments =
+      __ceiling_div(N, WorkGroupSize * iters_per_work_item);
+
+  sycl::buffer<BinType, 1> private_histograms(
```

Comment:
Is this API intended to support an async version? If not, it may be useful to experiment with USM instead of buffers here and see if there is any meaningful performance gain.

Reply:
In the context of this PR, we will not support an async version, but eventually in oneDPL this should be added to the async APIs but likely not at the start.  I'll look into this, thanks.